### PR TITLE
[awq] replace scale when we have GELU

### LINKED ISSUE #29421: Unused smoothing scales when loading AutoAWQ checkpoints
### System Info

- `transformers` version: 4.36.0
- Platform: Linux-5.15.0-1041-aws-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.20.3
- Safetensors version: 0.4.1
- Accelerate version: 0.25.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.1.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

Hi  @younesbelkada, I am loading an [AutoAWQ](https://github.com/casper-hansen/AutoAWQ/tree/main) checkpoint for the model [casperhansen/falcon-7b-awq](https://huggingface.co/casperhansen/falcon-7b-awq) and I get the following warning -

```python
import transformers
quant_path = "casperhansen/falcon-7b-awq"
model_hf = transformers.AutoModelForCausalLM.from_pretrained(quant_path)
```

Warning - 

```
Some weights of the model checkpoint at casperhansen/falcon-7b-awq were not used when initializing RWForCausalLM: ['transformer.h.4.mlp.act.scales', 'transformer.h.22.mlp.act.scales', 'transformer.h.14.mlp.act.scales', 'transformer.h.12.mlp.act.scales', 'transformer.h.5.mlp.act.scales', 'transformer.h.31.mlp.act.scales', 

'transformer.h.15.mlp.act.scales', 'transformer.h.9.mlp.act.scales', 'transformer.h.2.mlp.act.scales', 'transformer.h.13.mlp.act.scales', 'transformer.h.10.mlp.act.scales', 'transformer.h.7.mlp.act.scales', 

'transformer.h.3.mlp.act.scales', 'transformer.h.0.mlp.act.scales', 'transformer.h.16.mlp.act.scales', 'transformer.h.23.mlp.act.scales', 'transformer.h.6.mlp.act.scales', 'transformer.h.17.mlp.act.scales', 

'transformer.h.25.mlp.act.scales', 'transformer.h.29.mlp.act.scales', 'transformer.h.26.mlp.act.scales', 'transformer.h.28.mlp.act.scales', 'transformer.h.18.mlp.act.scales', 'transformer.h.20.mlp.act.scales', 

'transformer.h.21.mlp.act.scales', 'transformer.h.27.mlp.act.scales', 'transformer.h.8.mlp.act.scales', 'transformer.h.24.mlp.act.scales', 'transformer.h.30.mlp.act.scales', 'transformer.h.11.mlp.act.scales', 'transformer.h.19.mlp.act.scales', 'transformer.h.1.mlp.act.scales']
- This IS expected if you are initializing RWForCausalLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RWForCausalLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
```

This warning is shown when the activation in the MLP block is a GELU, but not for models with RELU (like [opt-awq](https://huggingface.co/casperhansen/opt-125m-awq)) or gated activations (like [llama2-awq](https://huggingface.co/TheBloke/Llama-2-7B-AWQ)). AutoAWQ handles this GELU case separately by creating a [`ScaledActivation` wrapper around GELU](https://github.com/casper-hansen/AutoAWQ/blob/main/awq/quantize/scale.py#L62-L65) that does the scaling inside the wrapper. 

This separate dedicated scaling seems to not be needed for RELU or GatedActivations because we can fuse this scaling operation with the weights of the first linear layer of the MLP -- 
- for RELU, because it is _kind of_ linear (and smoothing scales are positive), and 
- for GatedActivations, because we can scale one half of the first linear layer. 

There doesn't seem to be a convenient option like that for GELU (because of it being non-linear). 

**It's possible I am missing some implementation detail, but do you think this scaling needs to be added in when loading `AutoAWQ` checkpoints in `transformers` by reading the various `transformer.h.*.mlp.act.scales` values shown in the warning above?** 

When reading the same model (`casperhansen/falcon-7b-awq`) with `AutoAWQ` like below, we see the activation type is actually `ScaledActivation` -- 

```
from awq import AutoAWQForCausalLM
quant_path = "casperhansen/falcon-7b-awq"

model = AutoAWQForCausalLM.from_quantized(quant_path)

In [11]: print(model.model.transformer.blocks[0].mlp)
Out[11]:
MLP(
  (dense_h_to_4h): WQLinear_GEMM(in_features=4544, out_features=18176, bias=False, w_bit=4, group_size=64)
  (act): ScaledActivation(
    (act): GELU(approximate='none')
  )
  (dense_4h_to_h): WQLinear_GEMM(in_features=18176, out_features=4544, bias=False, w_bit=4, group_size=64)
)

```
OTOH, the `scales` inside the `ScaledActivation` are all `1`s for this model (effectively making no difference whether scaling is done or not) so maybe it was a deliberate decision to ignore the scales in the `transformers`'s AWQ integration? But I guess there is no guarantee that the scales with always be `1`s and we might have to handle this case separately?

```In [12]: model.model.transformer.blocks[0].mlp.act.scales
Out[12]:
Parameter containing:
tensor([1., 1., 1.,  ..., 1., 1., 1.], device='cuda:7', dtype=torch.float16,
       requires_grad=True)
```

Sorry for the long post, and please let me know if this belongs someplace else instead of a GH issue. Thanks!

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Transformers -- 

```python
import transformers
quant_path = "casperhansen/falcon-7b-awq"
model_hf = transformers.AutoModelForCausalLM.from_pretrained(quant_path)

print(model_hf.transformer.h[0].mlp.act)
```

Output -- 
```
GELU(approximate='none')
```

---

AutoAWQ -- 
```python
from awq import AutoAWQForCausalLM
quant_path = "casperhansen/falcon-7b-awq"
model = AutoAWQForCausalLM.from_quantized(quant_path)

print(model.model.transformer.blocks[0].mlp.act)
print(model.model.transformer.blocks[0].mlp.act.scales)
```

Output -- 
```
ScaledActivation(
  (act): GELU(approximate='none')
)

Parameter containing:
tensor([1., 1., 1.,  ..., 1., 1., 1.], device='cuda:7', dtype=torch.float16,
       requires_grad=True)

```

### Expected behavior

Expect the transformers AWQ integration to also have some form of smoothing/scaling for the second linear layer in the MLP block when GELU activation is used.

### LINKED ISSUE #30225: [BUG] Load StarCoder2 AWQ using Transformers
### System Info

Name: transformers
Version: 4.39.3

### Who can help?

@ArthurZucker 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Same to https://github.com/casper-hansen/AutoAWQ/issues/441
When loading starcoder2-AWQ using transformers, I received a confusing error:
```py
model = AutoModelForCausalLM.from_pretrained(
    "TechxGenus/starcoder2-3b-AWQ",
    torch_dtype=torch.float16,
    device_map="auto",
)
```
get:
> Some weights of the model checkpoint at TechxGenus/starcoder2-3b-AWQ were not used when initializing Starcoder2ForCausalLM: ['model.layers.0.mlp.act.scales', 'model.layers.1.mlp.act.scales', 'model.layers.10.mlp.act.scales', 'model.layers.11.mlp.act.scales', 'model.layers.12.mlp.act.scales', 'model.layers.13.mlp.act.scales', 'model.layers.14.mlp.act.scales', 'model.layers.15.mlp.act.scales', 'model.layers.16.mlp.act.scales', 'model.layers.17.mlp.act.scales', 'model.layers.18.mlp.act.scales', 'model.layers.19.mlp.act.scales', 'model.layers.2.mlp.act.scales', 'model.layers.20.mlp.act.scales', 'model.layers.21.mlp.act.scales', 'model.layers.22.mlp.act.scales', 'model.layers.23.mlp.act.scales', 'model.layers.24.mlp.act.scales', 'model.layers.25.mlp.act.scales', 'model.layers.26.mlp.act.scales', 'model.layers.27.mlp.act.scales', 'model.layers.28.mlp.act.scales', 'model.layers.29.mlp.act.scales', 'model.layers.3.mlp.act.scales', 'model.layers.4.mlp.act.scales', 'model.layers.5.mlp.act.scales', 'model.layers.6.mlp.act.scales', 'model.layers.7.mlp.act.scales', 'model.layers.8.mlp.act.scales', 'model.layers.9.mlp.act.scales']

And it doesn't output properly. There is no error when using `AutoAWQForCausalLM` or `vllm.LLM` to load it, and they work well.

### Expected behavior

This error is not generated and the output is correct

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do ? 
This PR replaces the scales by `ScaleActivation`  when we have GELU activation. Before this PR, the scales calculated during the quantization were not loaded properly. To have a quick fix that doesn't depend on each model, I just check if we have the following layers `layers = ["fc_in","dense_h_to_4h","up_proj","c_fc"]` in the module that contains the GELU. This is done in order to get the shape of the scales. We need that to be able to load the scales, otherwise we will get an error saying that the shapes don't match. 

Fixes https://github.com/huggingface/transformers/issues/29421
Fixes https://github.com/huggingface/transformers/issues/30225

### CODE REVIEW

File: src/transformers/integrations/awq.py [hunk 1]
```diff
@@ -145,6 +147,17 @@ def replace_with_awq_linear(
 
                 # Force requires grad to False to avoid unexpected errors
                 model._modules[name].requires_grad_(False)
+        if isinstance(module, nn.GELU) and name not in modules_to_not_convert:
+            # get the module just before applying act
+            # the easiest way is just checking if the mlp have these layers.
+            layers = ["fc_in", "dense_h_to_4h", "up_proj", "c_fc"]
```
Comment (amyeroberts) [inline src/transformers/integrations/awq.py:153]: This seems super brittle - let's do this in a more generalized way. In particular, this doesn't check these layers are used before applying the activation

File: src/transformers/integrations/awq.py [hunk 2]
```diff
@@ -145,6 +147,17 @@ def replace_with_awq_linear(
...
+            layers_detected = [l for l in layers if hasattr(model, l)]
+            if len(layers_detected) > 0:
+                size = getattr(model, layers_detected[0]).out_features
```
Comment (amyeroberts) [inline src/transformers/integrations/awq.py:156]: What if many layers are detected and they have different sizes?

File: src/transformers/integrations/awq.py [hunk 3]
```diff
@@ -145,6 +147,17 @@ def replace_with_awq_linear(
...
+                scale_like = torch.ones(size)
+                model._modules[name] = ScaledActivation(module, scale_like)
+            elif len(layers_detected) == 0:
+                logger.info("We were not able to replace the GELU layers. The model might not perform as expected.")
```
Comment (amyeroberts) [inline src/transformers/integrations/awq.py:160]: Hmmm, this doesn't seem ideal

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Two comments: 
* I'm worried that the current implementation will results in silent errors, as the logging is at the level of info. Hence, if some different module names than `["fc_in", "dense_h_to_4h", "up_proj", "c_fc"]` are used, this issue will not be solved and we'll rely on another user flagging this. 
* Tests should be added to make sure the replacement happens as expected

File: src/transformers/integrations/awq.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/integrations/awq.py]: Sounds good ! I decided to check for the model type and retrieve the layer before the activation.

File: src/transformers/integrations/awq.py (hunk 3 above)
Comment (SunMarc) [reply src/transformers/integrations/awq.py]: Fixed also !

File: src/transformers/integrations/awq.py [hunk 4]
```diff
@@ -145,6 +147,17 @@ def replace_with_awq_linear(
 
                 # Force requires grad to False to avoid unexpected errors
                 model._modules[name].requires_grad_(False)
+        if isinstance(module, nn.GELU) and name not in modules_to_not_convert:
```
[the remark this answers is not in the document]
Comment (SunMarc) [reply src/transformers/integrations/awq.py]: I've added the check for variants of GELU

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 

The only comment I'd make is to have a clearer name that `replace_scales`. `scales` isn't very descriptive - just from the method name, I don't really know what's happening, nor what a "scale" is

**Pull request merged**